Make image compression optional

kunimatsu-tri opened the pull request:
This PR provides an option for image compression in `ImageToLcmImageArrayT` to a user.

The current compression time is a few milliseconds for RGBA image and 30 to 40 milliseconds for Depth image. So, if you disable the image compressions in `ImageToLcmImageArrayT`, you can gain roughly 40 milliseconds run time improvement per an output port evaluation of `ImageToLcmImageArrayT`.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/robotlocomotion/drake/7318)
<!-- Reviewable:end -->

kunimatsu-tri commented:
+@thduynguyen feature review please. Thanks!

---

Review status: 0 of 3 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7318#-:-KxA29XRZHVbOAiquXh1:b-w57lpe)*
<!-- Sent from Reviewable.io -->

thduynguyen commented:
@kunimatsu-tri First pass review complete. PTAL.

---

Reviewed 3 of 3 files at r1.
Review status: all files reviewed at latest revision, 7 unresolved discussions.

---

*[drake/systems/sensors/image_to_lcm_image_array_t.h, line 14 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7318#-KxDiqWho6OymSz8Lov8:-KxDiqWho6OymSz8Lov9:b-xvrsnt) ([raw file](https://github.com/robotlocomotion/drake/blob/16ce935fe4cbfaf7356f7dc3442478fdae76f8e4/drake/systems/sensors/image_to_lcm_image_array_t.h#L14)):*
> ```objc
> namespace sensors {
> 
> /// An ImageToLcmImageArrayT takes as input a ImageBgra8U, ImageDepth32F and
> ```

BTW `an ImageBgra8U`

---

*[drake/systems/sensors/image_to_lcm_image_array_t.h, line 17 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7318#-KxDivsNRfUgyysIZ5n6:-KxDivsNRfUgyysIZ5n7:bkzm499) ([raw file](https://github.com/robotlocomotion/drake/blob/16ce935fe4cbfaf7356f7dc3442478fdae76f8e4/drake/systems/sensors/image_to_lcm_image_array_t.h#L17)):*
> ```objc
> /// ImageLabel16I. This system outputs an AbstractValue containing a
> /// `Value<robotlocomotion::image_array_t>` LCM message that defines an array
> /// of images (image_t). This message can then be sent to other processes that
> ```

BTW Should we explain the order (and meaning?) of the images in the LCM message in detail? It's not clear to me what the images in the array are.

---

*[drake/systems/sensors/image_to_lcm_image_array_t.h, line 26 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7318#-KxDlcC9uZ0RsARYq4oj:-KxDlcC9uZ0RsARYq4ok:bfb8d5h) ([raw file](https://github.com/robotlocomotion/drake/blob/16ce935fe4cbfaf7356f7dc3442478fdae76f8e4/drake/systems/sensors/image_to_lcm_image_array_t.h#L26)):*
> ```objc
>   DRAKE_NO_COPY_NO_MOVE_NO_ASSIGN(ImageToLcmImageArrayT)
> 
>   /// A %ImageToLcmImageArrayT constructor.
> ```

BTW `A` --> `An` ?

---

*[drake/systems/sensors/image_to_lcm_image_array_t.h, line 55 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7318#-KxDkgWvGLgfH8pjAZnF:-KxDkgWvGLgfH8pjAZnG:b-tfjili) ([raw file](https://github.com/robotlocomotion/drake/blob/16ce935fe4cbfaf7356f7dc3442478fdae76f8e4/drake/systems/sensors/image_to_lcm_image_array_t.h#L55)):*
> ```objc
>                       robotlocomotion::image_array_t* msg) const;
> 
>   int color_image_input_port_index_{};
> ```

BTW Sorry but I still have problems with this style. :) It's easy to be confused with an empty function definition on a quick glance. To me, `int color_image_input_port_index_ = 0;` is much clearer.

---

*[drake/systems/sensors/test/image_to_lcm_image_array_t_test.cc, line 59 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7318#-KxDnxX5FJjdMW1Q9Acu:-KxDnxX5FJjdMW1Q9Acv:b-tgvhuc) ([raw file](https://github.com/robotlocomotion/drake/blob/16ce935fe4cbfaf7356f7dc3442478fdae76f8e4/drake/systems/sensors/test/image_to_lcm_image_array_t_test.cc#L59)):*
> ```cpp
> 
>   auto Verify = [color_image, depth_image, label_image](
>                    const ImageToLcmImageArrayT& dut,
> ```

BTW What does `dut` mean?

---

*[drake/systems/sensors/test/image_to_lcm_image_array_t_test.cc, line 60 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7318#-KxDo63tz4IpAyzqw2t2:-KxDo63tz4IpAyzqw2t3:b-s37b7q) ([raw file](https://github.com/robotlocomotion/drake/blob/16ce935fe4cbfaf7356f7dc3442478fdae76f8e4/drake/systems/sensors/test/image_to_lcm_image_array_t_test.cc#L60)):*
> ```cpp
>   auto Verify = [color_image, depth_image, label_image](
>                    const ImageToLcmImageArrayT& dut,
>                    const robotlocomotion::image_array_t& output_image_array_t,
> ```

BTW I thought `_t` suffix usually denotes a type, but `output_image_array_t` is not.

---

*[drake/systems/sensors/test/image_to_lcm_image_array_t_test.cc, line 123 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7318#-KxDoZxYJot8m2qZcN5g:-KxDoZxYJot8m2qZcN5h:b-tgvhuc) ([raw file](https://github.com/robotlocomotion/drake/blob/16ce935fe4cbfaf7356f7dc3442478fdae76f8e4/drake/systems/sensors/test/image_to_lcm_image_array_t_test.cc#L123)):*
> ```cpp
>   };
> 
>   ImageToLcmImageArrayT dut_compressed(
> ```

BTW What does `dut` mean?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7318#-:-KxDorgDLSwzCCJ6WS_X:bm6u23)*
<!-- Sent from Reviewable.io -->

kunimatsu-tri commented:
Addressed all the comment. PTAL.

---

Review status: 1 of 3 files reviewed at latest revision, 7 unresolved discussions.

---

*[drake/systems/sensors/image_to_lcm_image_array_t.h, line 14 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7318#-KxDiqWho6OymSz8Lov8:-KxEO6ennvfVbsz_-0H2:bjagr70) ([raw file](https://github.com/robotlocomotion/drake/blob/16ce935fe4cbfaf7356f7dc3442478fdae76f8e4/drake/systems/sensors/image_to_lcm_image_array_t.h#L14)):*
<details><summary><i>Previously, thduynguyen (Duy-Nguyen Ta) wrote…</i></summary><blockquote>

BTW `an ImageBgra8U`
</blockquote></details>

Done.

Also, changed `ImageBgra8U` to `ImageRgba8U` since the doc was wrong.

---

*[drake/systems/sensors/image_to_lcm_image_array_t.h, line 17 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7318#-KxDivsNRfUgyysIZ5n6:-KxEOO1Vq1IKcCrPJTg9:b-qw8iax) ([raw file](https://github.com/robotlocomotion/drake/blob/16ce935fe4cbfaf7356f7dc3442478fdae76f8e4/drake/systems/sensors/image_to_lcm_image_array_t.h#L17)):*
<details><summary><i>Previously, thduynguyen (Duy-Nguyen Ta) wrote…</i></summary><blockquote>

BTW Should we explain the order (and meaning?) of the images in the LCM message in detail? It's not clear to me what the images in the array are.
</blockquote></details>

IMO,  the actual use case should be independent from the order.

I changed the unit test not to assume a particular order.

---

*[drake/systems/sensors/image_to_lcm_image_array_t.h, line 26 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7318#-KxDlcC9uZ0RsARYq4oj:-KxEP-b9rnXHEkVVSO6V:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/16ce935fe4cbfaf7356f7dc3442478fdae76f8e4/drake/systems/sensors/image_to_lcm_image_array_t.h#L26)):*
<details><summary><i>Previously, thduynguyen (Duy-Nguyen Ta) wrote…</i></summary><blockquote>

BTW `A` --> `An` ?
</blockquote></details>

Done.

---

*[drake/systems/sensors/image_to_lcm_image_array_t.h, line 55 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7318#-KxDkgWvGLgfH8pjAZnF:-KxEPRLOQjUGwTtYfa-C:b5nypkb) ([raw file](https://github.com/robotlocomotion/drake/blob/16ce935fe4cbfaf7356f7dc3442478fdae76f8e4/drake/systems/sensors/image_to_lcm_image_array_t.h#L55)):*
<details><summary><i>Previously, thduynguyen (Duy-Nguyen Ta) wrote…</i></summary><blockquote>

BTW Sorry but I still have problems with this style. :) It's easy to be confused with an empty function definition on a quick glance. To me, `int color_image_input_port_index_ = 0;` is much clearer.
</blockquote></details>

This is widely used convention in Drake, so I just put `-1` to be clearer (avoided using `0` since it looks something meaningful, but it is not.)

If you really want to discourage using `{}`, I recommend you to propose the new convention to all the Drake developers either via Slack or an Github issue.

---

*[drake/systems/sensors/test/image_to_lcm_image_array_t_test.cc, line 59 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7318#-KxDnxX5FJjdMW1Q9Acu:-KxEEemv9mfgtHerT-5_:b-ncoavr) ([raw file](https://github.com/robotlocomotion/drake/blob/16ce935fe4cbfaf7356f7dc3442478fdae76f8e4/drake/systems/sensors/test/image_to_lcm_image_array_t_test.cc#L59)):*
<details><summary><i>Previously, thduynguyen (Duy-Nguyen Ta) wrote…</i></summary><blockquote>

BTW What does `dut` mean?
</blockquote></details>

I think it's Device Under Test.  It's a widely used naming convention in Drake's unit test.

---

*[drake/systems/sensors/test/image_to_lcm_image_array_t_test.cc, line 60 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7318#-KxDo63tz4IpAyzqw2t2:-KxET8HcQNEnRbx1YMyh:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/16ce935fe4cbfaf7356f7dc3442478fdae76f8e4/drake/systems/sensors/test/image_to_lcm_image_array_t_test.cc#L60)):*
<details><summary><i>Previously, thduynguyen (Duy-Nguyen Ta) wrote…</i></summary><blockquote>

BTW I thought `_t` suffix usually denotes a type, but `output_image_array_t` is not.
</blockquote></details>

Done.

---

*[drake/systems/sensors/test/image_to_lcm_image_array_t_test.cc, line 123 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7318#-KxDoZxYJot8m2qZcN5g:-KxEEhAqI-z7WewWSdli:b-pvpk9c) ([raw file](https://github.com/robotlocomotion/drake/blob/16ce935fe4cbfaf7356f7dc3442478fdae76f8e4/drake/systems/sensors/test/image_to_lcm_image_array_t_test.cc#L123)):*
<details><summary><i>Previously, thduynguyen (Duy-Nguyen Ta) wrote…</i></summary><blockquote>

BTW What does `dut` mean?
</blockquote></details>

See the earlier comment.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7318#-:-KxEVaIxTvwJknqNr5-1:b91tz3i)*
<!-- Sent from Reviewable.io -->

thduynguyen commented:
Two more comments. :)

---

Reviewed 2 of 2 files at r2.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[drake/systems/sensors/image_to_lcm_image_array_t.h, line 17 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7318#-KxDivsNRfUgyysIZ5n6:-KxEfb7sSHr9fIyRJwwE:bgd837f) ([raw file](https://github.com/robotlocomotion/drake/blob/16ce935fe4cbfaf7356f7dc3442478fdae76f8e4/drake/systems/sensors/image_to_lcm_image_array_t.h#L17)):*
<details><summary><i>Previously, kunimatsu-tri (Kunimatsu Hashimoto) wrote…</i></summary><blockquote>

IMO,  the actual use case should be independent from the order.

I changed the unit test not to assume a particular order.
</blockquote></details>

BTW I would add to the comment that we assume no particular order of the output images and users have to check the semantic of those images via `image.pixel_format` (as done in the unit test).  This would help set up the user's expectation for the output of this system.

---

*[drake/systems/sensors/image_to_lcm_image_array_t.h, line 20 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7318#-KxEeknzWmMGhUu63NxO:-KxEeknzWmMGhUu63NxP:bwkowau) ([raw file](https://github.com/robotlocomotion/drake/blob/a5d5b16a770de84907d9f22805bc4660e5f70f34/drake/systems/sensors/image_to_lcm_image_array_t.h#L20)):*
> ```objc
> /// sbscribe it using LcmPublisherSystem.
> // TODO(kunimatsu-tri) Instead of assuming fixed pixel types for the input
> // ports, e.g. ImageBgra8U, change the interface to be able to handle arbitrary
> ```

BTW Bgra->Rgba?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7318#-:-KxEh55aI2WM0mSqJSjP:buiub8n)*
<!-- Sent from Reviewable.io -->

kunimatsu-tri commented:
Addressed those two comments, PTAL.

---

Review status: 2 of 3 files reviewed at latest revision, 2 unresolved discussions.

---

*[drake/systems/sensors/image_to_lcm_image_array_t.h, line 17 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7318#-KxDivsNRfUgyysIZ5n6:-KxErncBX23Kv_FuVC6P:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/16ce935fe4cbfaf7356f7dc3442478fdae76f8e4/drake/systems/sensors/image_to_lcm_image_array_t.h#L17)):*
<details><summary><i>Previously, thduynguyen (Duy-Nguyen Ta) wrote…</i></summary><blockquote>

BTW I would add to the comment that we assume no particular order of the output images and users have to check the semantic of those images via `image.pixel_format` (as done in the unit test).  This would help set up the user's expectation for the output of this system.
</blockquote></details>

Done.

---

*[drake/systems/sensors/image_to_lcm_image_array_t.h, line 20 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7318#-KxEeknzWmMGhUu63NxO:-KxEpRhCCtoGM3MBRw8c:b4acftw) ([raw file](https://github.com/robotlocomotion/drake/blob/a5d5b16a770de84907d9f22805bc4660e5f70f34/drake/systems/sensors/image_to_lcm_image_array_t.h#L20)):*
<details><summary><i>Previously, thduynguyen (Duy-Nguyen Ta) wrote…</i></summary><blockquote>

BTW Bgra->Rgba?
</blockquote></details>

Thanks, a good catch!

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7318#-:-KxEru95ZcfpbE21-OlS:b-5g6nia)*
<!-- Sent from Reviewable.io -->

thduynguyen commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 1 of 1 files at r3.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7318#-:-KxEy43wmNIsdaUhXW5U:bnfp4nl)*
<!-- Sent from Reviewable.io -->

kunimatsu-tri commented:
+@sammy-tri platform review please. Thanks!

---







*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7318#-:-KxFEAmvoxKeIq0Yrhjq:b3j5bdc)*
<!-- Sent from Reviewable.io -->

sammy-tri commented:
+(status: curate commits before merging) <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 1 of 2 files at r2, 2 of 2 files at r4.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7318#-:-KxFRwklrM32cVMkKJtu:bsf2vkf)*
<!-- Sent from Reviewable.io -->

kunimatsu-tri commented:
Thanks both @thduynguyen and @sammy-tri!

---







*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7318#-:-KxFSIfBJ_-jzhsXUv7I:b8934eu)*
<!-- Sent from Reviewable.io -->
